### ISSUE TITLE
Jump step Two-point difference is throwing warnings in its tests

### ISSUE THREAD

**Issue by jdavies-st** [spacetelescope/jwst#1853]
When running the tests in the jump step two-point difference module, the same RuntimeWarning gets thrown for many of the tests:
```python
(conda_dev3) solo> cd jwst/jump/tests
(conda_dev3) solo> pytest -n 12 .
===================================================================== test session starts ======================================================================
platform darwin -- Python 3.5.5, pytest-3.5.0, py-1.5.3, pluggy-0.6.0
rootdir: /Users/jdavies/dev/jwst, inifile: setup.cfg
plugins: xdist-1.22.1, forked-0.2, hypothesis-3.52.0
gw0 [31] / gw1 [31] / gw2 [31] / gw3 [31] / gw4 [31] / gw5 [31] / gw6 [31] / gw7 [31] / gw8 [31] / gw9 [31] / gw10 [31] / gw11 [31]
scheduling tests via LoadScheduling
.......x..............x........                                                                                                                          [100%]
======================================================================= warnings summary =======================================================================
jwst/jump/tests/test_twopoint_difference.py::test_4grps_cr2_NoFlux
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_4grps_twocrs_2nd_4th
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_6grps_negative_differences_zeromedian
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_5grps_twocrs_2nd_5thbig_nframes2
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_6grps_noCR
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_7grps_1CR
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_6grps_1CR
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_10grps_cr2_gt3sigma
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_10grps_twocrs_10percenthit
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_6grps_twocrs_twopixels_nframes2
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_10grps_twocrs_2nd_8th_big
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_5grps_cr2_negslope
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_6grps_twocrs_2nd_5th_nframes2
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_6grps_Satat6_CRat1_flagadjpixels
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_6grps_twocrs_2nd_5th
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_6grps_Satat6_CRat1
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_10grps_Satat8_CRsat3and6
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_10grps_cr2_3sigma_2frames_noCR
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_5grps_noCR
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

jwst/jump/tests/test_twopoint_difference.py::test_10grps_cr2_gt3sigma_2frames
  /Users/jdavies/dev/jwst/jwst/jump/twopoint_difference.py:83: RuntimeWarning: invalid value encountered in sqrt
    poisson_noise = np.sqrt(np.abs(med_diffs))

-- Docs: http://doc.pytest.org/en/latest/warnings.html
====================================================== 29 passed, 2 xfailed, 20 warnings in 56.74 seconds ======================================================
```
But it is not clear why, as I believe the only reason `np.sqrt` would encounter an invalid value is if it is negative, but the code applies `np.abs` first, so I'm at a loss.

**Comment by jdavies-st** [comment]
So it turns out this warning only gets thrown when one uses `float32` and the array is sufficiently large and all zeros.  I've [reported an issue](https://github.com/numpy/numpy/issues/10838) to the numpy folks.

**Comment by jdavies-st** [comment]
Well, it turns it it is not a `numpy` issue.  It only seems to happen on Mac OS running `numpy` from `conda`.

@rendinam, you noticed that there was some square-root-of-almost-zero problems in some other regression tests as well?  Any more details?

Any idea where I might file this issue to the Anaconda folks?  Who is responsible for making the recipes that build the binaries that eventually we pull down as `numpy` via `conda`?  @jhunkeler?

**Comment by jdavies-st** [comment]
Here is a simple example of what is happening.

For a small 3x3 array of zeros, `sqrt()` does not encounter any negative numbers:
```
In [3]: a = np.zeros((3, 3), dtype=np.float32)

In [4]: np.sqrt(a)
Out[4]: 
array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]], dtype=float32)
```
But for a sufficiently large array of zeros, there's some negative numbers in there:
```
In [5]: a = np.zeros((100, 100), dtype=np.float32)

In [6]: np.sqrt(a)
/Users/jdavies/anaconda/envs/conda_dev3/bin/ipython:1: RuntimeWarning: invalid value encountered in sqrt
  #!/Users/jdavies/anaconda/envs/conda_dev3/bin/python
Out[6]: 
array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

In [7]: np.__version__
Out[7]: '1.14.2'
```

**Comment by philhodge** [comment]
Perhaps you should post a note on Numpy-discussion.

**Comment by rendinam** [comment]
The problem I noticed wasn't a square root issue specifically, but a correlation between unexpected test failures in calcos stemming from tiny numerical discrepancies and a change in the numpy package present in the testing environment. Also, this happened when using numpy 1.13.3.

The environment changes that happened between passing tests and failing ones (3/27/2018 - 3/28/2018):
```
numpy:   1.13.3-py35ha266831_3    |        numpy:         1.13.3-py35hdbf6ddf_4
                                  >     mkl_fft:          1.0.1-py35h3010b51_0
                                  >     mkl_random:       1.0.1-py35h629b387_0
```

This indicates that numpy 1.13.3 was rebuilt, presumably to accommodate the addition of the mkl packages that were added.

numpy 1.14.2 did appear to get rebuilt and a new package of the same version published on 3/23/2018.

This is circumstantial, but multiple numerical issues popping up at the same time in various contexts alongside numpy conda package changes is certainly suspicious.

**Comment by jdavies-st** [comment]
Thanks @rendinam.  I did notice the changes to how it is compiled with MKL.  Compared a 1.12.1 version of Numpy in an old conda environment to the one that is currently failing in 1.14.2 in the `site-packages/numpy/__config__.py` file:

```
--- numpy-1.12.1
> blas_mkl_info={'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'libraries': ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread'], 'include_dirs': ['/Users/jdavies/anaconda/envs/dev/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/dev/lib']}
> blas_opt_info={'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'libraries': ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread'], 'include_dirs': ['/Users/jdavies/anaconda/envs/dev/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/dev/lib']}
> lapack_opt_info={'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'libraries': ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread'], 'include_dirs': ['/Users/jdavies/anaconda/envs/dev/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/dev/lib']}
> lapack_mkl_info={'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'libraries': ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread'], 'include_dirs': ['/Users/jdavies/anaconda/envs/dev/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/dev/lib']}
--- numpy-1.14.2
< mkl_info={'libraries': ['mkl_rt', 'pthread'], 'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'include_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/lib']}
< blas_mkl_info={'libraries': ['mkl_rt', 'pthread'], 'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'include_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/lib']}
< lapack_mkl_info={'libraries': ['mkl_rt', 'pthread'], 'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'include_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/lib']}
< blas_opt_info={'libraries': ['mkl_rt', 'pthread'], 'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'include_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/lib']}
< lapack_opt_info={'libraries': ['mkl_rt', 'pthread'], 'define_macros': [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)], 'include_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/include'], 'library_dirs': ['/Users/jdavies/anaconda/envs/conda_dev3/lib']}
```

**Comment by jdavies-st** [comment]
This issue will be tracked over at Anaconda
https://github.com/ContinuumIO/anaconda-issues/issues/9129

**Comment by jdavies-st** [comment]
One thing I notice is that in the anaconda repo where we pull our MacOS build:

https://repo.anaconda.com/pkgs/main/osx-64/

There are 4 numpy 1.14.2 for python 3.6 builds:
```
numpy-1.14.2-py36_nomklh9797aa9_0.tar.bz2 | 3.9 MB | 2018-03-12 22:15:35 UTC | f3f832cdf90c673941d1469290c68b8a
numpy-1.14.2-py36_nomklh9797aa9_1.tar.bz2 | 3.9 MB | 2018-03-23 14:20:29 UTC | ecd14781f6480c7350f440374e63827c
numpy-1.14.2-py36ha9ae307_0.tar.bz2 | 3.9 MB | 2018-03-12 22:13:12 UTC | 4b321d4a263cd6f60d4a9be7749473c1
numpy-1.14.2-py36ha9ae307_1.tar.bz2 | 4.0 MB | 2018-03-23 14:23:02 UTC | aad7a39434d427467512c56962406940
```
We're pulling the last one, which is build 1 of numpy 1.14.2 with MKL acceleration.  I suspect the MKL acceleration is what is causing this issue.  Can I (or we) easily pull in the previous build (build 0) or one without MKL acceleration?  Even just to do a little test?

**Comment by jdavies-st** [comment]
So according to the Anaconda folks, this is a problem with the way Intel's Vector Math Library reports errors/warnings see that issue above for the details.

The bottom line though is that the zeros in large numpy arrays are in fact zeros, but the warning is being triggered anyway.  So I don't think this is producing corrupt or incorrect data.  It is only producing incorrect warnings.

I don't think we need to do anything on our end to fix our packaging or modify our use of numpy.

**Comment by bernie-simon** [comment]
You can silence the error messages with

    with np.errstate(invalid='ignore'):
         b = np.sqrt(a)

**Comment by jdavies-st** [comment]
This has been fixed in `mkl 2019.0`.  Closing.
